Update disk resize to work with `lvms-storage-operator`/topolvm

anjannath opened the pull request:
fixes #4117

needs https://github.com/crc-org/snc/pull/877

cfergeau commented on pkg/crc/config/settings.go:
```diff
@@ -37,6 +37,7 @@ const (
 	EmergencyLogin           = "enable-emergency-login"
 	PersistentVolumeSize     = "persistent-volume-size"
 	EnableBundleQuayFallback = "enable-bundle-quay-fallback"
+	SecondaryDiskSize        = "secondary-disk-size"
```
Isn't this similar to `PersistentVolumeSize`?

anjannath replied:
yes practically they are the same, one difference is that `secondary-disk-size` is at the VM level similar to `cpus`,`memory` and `persistent-volume-size` is at the openshift cluster level

cfergeau replied:
Do we need to expose this difference to the user?

anjannath replied:
nope, going to use the same config key `persistent-volume-size` and remove the newly added one

cfergeau commented on pkg/drivers/vfkit/driver_darwin.go:
```diff
@@ -135,6 +143,14 @@ func (d *Driver) Create() error {
+		if crcos.FileExists(d.getSecondaryDiskPath()) {
+			return nil
+		}
+		_, err := diskfs.Create(d.getSecondaryDiskPath(), int64(d.SecondaryDiskSize), diskfs.Raw, diskfs.SectorSizeDefault)
```
Is `diskfs` required? Could this use `os.Truncate` directly to create an empty disk image?

anjannath replied:
yes, this works, tried using `os.Truncate`, we have to first create an empty file and then change the size using `os.Truncate`

gbraad commented on cmd/crc/cmd/start.go:
```diff
@@ -79,6 +79,7 @@ func runStart(ctx context.Context) (*types.StartResult, error) {
 		IngressHTTPPort:   config.Get(crcConfig.IngressHTTPPort).AsUInt(),
 		IngressHTTPSPort:  config.Get(crcConfig.IngressHTTPSPort).AsUInt(),
 		EnableSharedDirs:  config.Get(crcConfig.EnableSharedDirs).AsBool(),
+		SecondaryDiskSize: config.Get(crcConfig.SecondaryDiskSize).AsInt(),
```
Here we assume only a second disk could exist. But what about `Disks` ?

anjannath replied:
to make it work with more disks, we'd have to create a separate `DiskConfig` type maybe.. containing the path and the size then add that to the drivers, i went with this simpler approach to first take care of the requirements for installing topolvm

anjannath commented:
this PR is not needed anymore, the snc PRs:
1. https://github.com/crc-org/snc/pull/869
2. https://github.com/crc-org/snc/pull/869
will take care of all the steps needed to install topolvm entirely on snc

anjannath committed:
use label to find the root partition in ocp preset

after using topolvm as storage provisioner there's
two partitions of type xfs and its ambiguos  which
is the root partition but the root partition  also
has a label 'root' in case of the ocp bundle

anjannath committed:
update disk resize to work with new disk layout

to be able to use topolvm we need a separate parition this new
partition is added at the end of the disk and used by  topolvm
so the new disk layout is |efi|boot|root|pv|

earlier the root parition could be easily extended after  increasing
the disk size as there was no patition after it, with the new layout
after increasing the disk size we have to move the "pv" partition to
create free space after the "root" partition to increase its size

this adds code to move topolvm partition and restart the VM which is
needed to re-read the changed partition table, after moving the part
the root partition can be grown

praveenkumar commented:
/wip

praveenkumar commented:
/hold

anjannath committed:
config: make persistent-volume-size applicable for ocp preset

this allows users to also configure the persistent volume's partition
size for the ocp preset, in case of ocp preset the default size of the
persistent volume partition (topolvm partition) is 3G

anjannath committed:
refactor: move function related to paritions into separate file

this makes it a bit easier to navigate the start.go file by moving
some of the disk size and partition modification functions into  a
separate disks.go file in the same machine package

cfergeau commented on pkg/crc/machine/start.go:
```diff
@@ -151,11 +151,11 @@ func growRootFileSystem(sshRunner *crcssh.Runner, preset crcPreset.Preset, persi
+		cmd = "/usr/sbin/blkid -t TYPE=LVM2_member -o device"
 	}
-	part, _, err := sshRunner.RunPrivileged("Get device id", "/usr/sbin/blkid", "-t", fmt.Sprintf("TYPE=%s", diskType), "-o", "device")
+	part, _, err := sshRunner.RunPrivileged("Get device id", cmd)
```
While`sshRunner.RunPrivileged(cmd)` will work, the `RunPrivileged` interface expects split args (ie a `[]string`) as other implementations won't work with a single string containing all the args separated with spaces.

cfergeau replied:
Might be better to avoid the duplication by using `query := "--label root"` and `query = "-t TYPE=LVM2_member"` and then hardcoding `/usr/sbin/blkid` and `-o device` in a single place.

anjannath replied:
updated the commit to use the above suggestion!

cfergeau commented on pkg/crc/machine/start.go:
```diff
@@ -126,156 +125,6 @@ func restartHost(ctx context.Context, vm *virtualMachine, sshRunner *crcssh.Runn
-	}
-
-	// with '/dev/[sv]da4' as input, run 'growpart /dev/[sv]da 4'
-	if _, _, err := sshRunner.RunPrivileged(fmt.Sprintf("Growing %s partition", rootPart), "/usr/bin/growpart", rootPart[:len("/dev/.da")], rootPart[len("/dev/.da"):]); err != nil {
```
The corresponding code in the new file has become
```
	if err := growPersistentVolume(sshRunner, startConfig.Preset, startConfig.PersistentVolumeSize); err != nil {
		return fmt.Errorf("Unable to grow persistent volume partition: %w", err)
	}

	// with '/dev/[sv]da4' as input, run 'growpart /dev/[sv]da 4'
	if err := growPartition(sshRunner, rootPart); err != nil {
		return nil
	}
```

While the addition of a helper is a good thing, this should not be hidden in a commit saying it's only moving code.
Fwiw, a helper for `move topolvm partition to end of disk` could also be useful to make `growRootFilesystem` easier to read.

cfergeau reviewed (commented):
I'm a bit confused by ' update disk resize to work with new disk layout ', could you detail more in the commit log why we need to move the topolvm partition (ie what fails if we don't do this)? What is the order before, what is the order after?  Could this be done proactively in snc? And can we resize both the root partition and the PVs without any issue?

cfergeau commented:
> this PR is not needed anymore, the snc PRs:

I assume this is no longer accurate as you updated the PR?

anjannath committed:
remove hardcoded topolvm partition name

this uses `blkid -t PARTLABEL=topolvm` to get the topolvm
partition name and introduces a helper 'runBlkidQuery' to
run different forms of the `blkid` command

anjannath commented:
> I'm a bit confused by ' update disk resize to work with new disk layout ', could you detail more in the commit log why we need to move the topolvm partition (ie what fails if we don't do this)? What is the order before, what is the order after? Could this be done proactively in snc? And can we resize both the root partition and the PVs without any issue?

i've updated the commit message.. but mentioning here as well: after adding a new partition at the end of disk, the partition layout is `|efi|boot|root|topolvm|`, now if we want to increase the size of the root partition we have to first move the `topolvm` partition to create the required free space after the `root` partition, earlier the `root` partition was the last partition on the disk and after increasing the disk size the additional free space was added after it, which is not the case anymore because the `topolvm` partition is there now

cfergeau commented:
> > I'm a bit confused by ' update disk resize to work with new disk layout ', could you detail more in the commit log why we need to move the topolvm partition (ie what fails if we don't do this)? What is the order before, what is the order after? Could this be done proactively in snc? And can we resize both the root partition and the PVs without any issue?
> 
> i've updated the commit message.. but mentioning here as well: after adding a new partition at the end of disk, the partition layout is `|efi|boot|root|topolvm|`, now if we want to increase the size of the root partition we have to first move the `topolvm` partition to create the required free space after the `root` partition, earlier the `root` partition was the last partition on the disk and after increasing the disk size the additional free space was added after it, which is not the case anymore because the `topolvm` partition is there now

Ah ok, we need to go from `|efi|boot|root|topolvm|free space|` to `|efi|boot|root|free space|topolvm|`. The `topolvm` partition is always the last one, but its end also has to be aligned with the end of the disk so that the freed space comes before rather than after it.

praveenkumar commented on pkg/crc/constants/constants.go:
```diff
@@ -223,6 +221,18 @@ func GetDefaultMemory(preset crcpreset.Preset) int {
+		return 15
+	default:
+		// should not be reached
+		return 3
```
Will it be enough? because registry also going to use the pvc which is going to part of this partition and if a user mirror an image to openshift or even build a image from openshift then this space is going to be filled very fast :(

anjannath replied:
right, we might have to increase the default disk-size, but there's very limited free space available, and currently that space is  shared for both image-registry and host container image storage, with this since we have to create a separate partition this becomes more apparent:

```
[core@crc ~]$ df -h /dev/vda4
Filesystem      Size  Used Avail Use% Mounted on
/dev/vda4        31G   23G  7.6G  76% /
```

We only have 7.6G free space, of which now `4.6G` is for host container images and `3G` for PVs

anjannath replied:
maybe we can increase the default `disk-size` to 40 and default `persistent-volume-size` to 7, these are just arbitrary numbers, but it'd leave around 7gb free space on the root partition (closer to the current situation)

(the bundle sizes are also increasing which would mean less free space for workloads, right? i saw the `4.16.3` bundle from CI in 400MB larger than the 4.16.0 bundle)

cfergeau replied:
For what it's worth, increasing the guest disk size should not increase the bundle size nor the used disk space after `crc start` as long as the additional space is empty.

praveenkumar commented on pkg/crc/machine/start.go:
```diff
@@ -112,12 +112,57 @@ func (client *client) updateVMConfig(startConfig types.StartConfig, vm *virtualM
+		return errors.Wrap(err, "Failed to connect to the CRC VM with SSH -- virtual machine might be unreachable")
+	}
+	return nil
+}
```
This restart means add more time to startup even it takes ~1 mins.

anjannath replied:
yes, its less than 1 min, but that's on linux, on windows i expect it to be even longer than 1 minute

cfergeau replied:
Is the restart required? If I run the `sfdisk` command manually, I get
```
The partition table has been altered.
Calling ioctl() to re-read partition table.
Re-reading the partition table failed.: Device or resource busy
The kernel still uses the old table. The new table will be used at the next reboot or after you run partprobe(8) or partx(8).
Syncing disks.
```
Running `partprobe` could be enough? 

I'd really prefer to avoid restarts of the VM/... during a `crc start`, this feels like a complication which will cause us problems later on.

anjannath replied:
running `partprobe` was not enough, it needed the reboot, yeah i agree that this makes it complicated and now it works because the VM ip doesn't change after a reboot

cfergeau commented on pkg/crc/machine/start.go:
```diff
@@ -150,6 +188,35 @@ func growRootFileSystem(sshRunner *crcssh.Runner, preset crcPreset.Preset, persi
+
+	if preset == crcPreset.OpenShift {
+		pvPartition := "/dev/vda5"
+		if _, _, err := sshRunner.RunPrivileged(fmt.Sprintf("Growing %s partition", pvPartition), "/usr/bin/growpart", pvPartition[:len("/dev/.da")], pvPartition[len("/dev/.da"):]); err != nil {
```
The `add helper to grow root partition` should come before this commit, this way you can reuse it here instead of duplicating already existing code.

cfergeau commented on pkg/crc/machine/start.go:
```diff
@@ -112,12 +112,57 @@ func (client *client) updateVMConfig(startConfig types.StartConfig, vm *virtualM
+			}
+			if exitErr.ExitStatus() != 1 {
+				return err
+			}
```
Do you need to ignore errors when the command exited with 1 as its exit code? Or is it a copy and paste error?

anjannath replied:
yes we need to ignore it, after moving the partition it returns exit code 1 (i think because of `Re-reading the partition table failed.: Device or resource busy` but couldn't find anything about exit codes in `sfdisk` man page)

cfergeau commented on pkg/crc/machine/start.go:
```diff
@@ -112,12 +112,57 @@ func (client *client) updateVMConfig(startConfig types.StartConfig, vm *virtualM
+func growRootFileSystem(ctx context.Context, startConfig types.StartConfig, vm *virtualMachine, sshRunner *crcssh.Runner) error {
+	if startConfig.Preset == crcPreset.OpenShift {
+		sizeToMove := ocpGetPVShiftSize(startConfig.DiskSize, startConfig.PersistentVolumeSize)
+		_, _, err := sshRunner.RunPrivileged("move topolvm partition to end of disk", fmt.Sprintf("echo '+%dG,' | sudo sfdisk --move-data /dev/vda -N 5 --force", sizeToMove))
```
I would squash `add helper to move topolvm partition` with this commit, less code to review overall :)

cfergeau commented on pkg/crc/machine/start.go:
```diff
@@ -112,12 +112,57 @@ func (client *client) updateVMConfig(startConfig types.StartConfig, vm *virtualM
+	return diskSize - constants.DefaultDiskSize
+}
+
+func growRootFileSystem(ctx context.Context, startConfig types.StartConfig, vm *virtualMachine, sshRunner *crcssh.Runner) error {
```
This is already the case before this commit, but `growRootFileSystem` is badly named, it grows the root filesystem, and the PVs filesystem. `growFilesystems` would be a better name for it.

cfergeau commented on pkg/crc/machine/start.go:
```diff
@@ -112,12 +112,57 @@ func (client *client) updateVMConfig(startConfig types.StartConfig, vm *virtualM
+	if pvSize > defaultPvSize {
+		return (diskSize - constants.DefaultDiskSize) - (pvSize - defaultPvSize)
+	}
+	return diskSize - constants.DefaultDiskSize
```
I'd recommend making use of `github.com/containers/common/pkg/strongunits` here (and in more places in our codebase :). At the very least, add a GiB suffix to the function name so that there is no confusion about what we are manipulating.

cfergeau replied:
I'd favour calculating this from `sfdisk` output though, if the disk layout is in an unexpected state, these calculations could break as they rely on external values, not on the actual disk state. What we want to do is to align the last partition to the end of the disk, so we can do it without any knowledge of the default disk size, the disk size stored in json metadata ,...

anjannath replied:
> What we want to do is to align the last partition to the end of the disk, so we can do it without any knowledge of the default disk size, the disk size stored in json metadata

that is not the case always, sometime we want to move the partition not to the very end and leave space to increase the PV partition size

suppose `disk-size` is `40` and `persistent-volume-size` is `5` (all in GB) then after increasing the entire disk to 40GB from 31GB the last partition needs to be moved by 7GB (not 9GB) leaving 2GB of space at the end so the PV partition can grow to become 5GB (default size is 3GB)

cfergeau replied:
We could grow the partition before moving it since it's at the end, or we can take into account its new size desired size when moving it, this is fine as this is data we cannot know from the current layout.

My impression from the current is that we determine the offset we need to move the partition without ever looking at the actual on disk layout, but only using hardcoded constants and crc configuration values. If the disk layout (unexpectedly) goes out of sync with this, bad things will happen I think?

anjannath replied:
i think we can replace `constant.DefaultDiskSize` and `defaultPvSize` with the current disk size and current pv partition size fetching it with `sfdisk`, but i think it'll be an edge case if the config values do not reflect the actual values

anjannath replied:
added https://github.com/crc-org/crc/pull/4114/commits/d129b861633b41aadc73767b7b43959ccd0f94fb to address this 

i'll try to incorporate this into the original commit, but since the functions gets moved to another file, its taking a bit of time to do this

anjannath committed:
add helper to grow root partition

cfergeau commented on pkg/crc/machine/disks.go:
```diff
@@ -39,7 +39,12 @@ func ocpGetPVShiftSize(diskSize int, pvSize int) int {
+		return err
+	}
+	_, _, err = sshRunner.RunPrivileged("move topolvm partition to end of disk",
+		fmt.Sprintf("echo '+%dG,' | sudo sfdisk --move-data /dev/vda -N %s --force", shiftSize, pvPartition[len("/dev/.da"):]))
```
If you use `len("/dev/.da")`, it might be worth also making the `/dev/vda` more generic? Iirc it's `/dev/sda` on Windows?

cfergeau commented on pkg/crc/machine/disks.go:
```diff
@@ -125,6 +133,14 @@ func getrootPartition(sshRunner *crcssh.Runner, preset crcPreset.Preset) (string
+	return getDeviceID(sshRunner, "-t PARTLABEL=topolvm")
+}
+
+func getDeviceID(sshRunner *crcssh.Runner, query string) (string, error) {
```
I find the api to this function a bit odd, this `query` string is quite magic. Might be better to name it `runBlkidQuery` or such?

cfergeau commented on pkg/crc/machine/disks.go:
```diff
@@ -125,6 +133,14 @@ func getrootPartition(sshRunner *crcssh.Runner, preset crcPreset.Preset) (string
+}
+
+func getTopolvmPartition(sshRunner *crcssh.Runner) (string, error) {
+	return getDeviceID(sshRunner, "-t PARTLABEL=topolvm")
```
I'd also rework `getDeviceID` to accept a `...string` rather than just a `string` as the correct (in theory) way of calling it is `getDeviceID(sshRunner, "-t", "PARTLABEL=topolvm")`

I made this change on top of yours:

```
diff --git a/pkg/crc/machine/disks.go b/pkg/crc/machine/disks.go
index 4283d1501..2ea169de0 100644
--- a/pkg/crc/machine/disks.go
+++ b/pkg/crc/machine/disks.go
@@ -129,19 +131,20 @@ func growPersistentVolume(sshRunner *crcssh.Runner, preset crcPreset.Preset, per
 }
 
 func getrootPartition(sshRunner *crcssh.Runner, preset crcPreset.Preset) (string, error) {
-	query := "--label root"
 	if preset == crcPreset.Microshift {
-		query = "-t TYPE=LVM2_member"
+		return getDeviceID(sshRunner, "-t", "TYPE=LVM2_member")
 	}
-	return getDeviceID(sshRunner, query)
+	return getDeviceID(sshRunner, "--label", "root")
 }
 
 func getTopolvmPartition(sshRunner *crcssh.Runner) (string, error) {
-	return getDeviceID(sshRunner, "-t PARTLABEL=topolvm")
+	return getDeviceID(sshRunner, "-t", "PARTLABEL=topolvm")
 }
 
-func getDeviceID(sshRunner *crcssh.Runner, query string) (string, error) {
-	part, _, err := sshRunner.RunPrivileged("Get device id", "/usr/sbin/blkid", query, "-o", "device")
+func getDeviceID(sshRunner *crcssh.Runner, query ...string) (string, error) {
+	cmd := []string{"/usr/sbin/blkid", "-o", "device"}
+	cmd = append(cmd, query...)
+	part, _, err := sshRunner.RunPrivileged("Get device id", cmd...)
 	if err != nil {
 		return "", err
 	}
```

anjannath replied:
thanks, i applied this patch on the commit, and rename the function to `runBlkidQuery`

anjannath committed:
refactor: rename growRootFileSystem to growFileSystem

since it does more than growing only the root file system
it additionally grows the partition used by  topolvm/lvms
operator

anjannath committed:
disk-resize: use actual partition sizes to calculate pv shift size

instead of relying on the config values and the constants for default
disk-size and pv partition size, this fetches the size of disk,  root
and pv partition using `lsblk` and uses these values to calculate the
size by which the pv partition needs to be shifted to apply requested
new values for 'disk-size' and 'persistent-volume-size' configs

it uses the 'strongunits' package from containers/common for bytes to
GiB conversions and its types to deal with storage sizes

anjannath commented:
seems this is not a good return on investment, closing it, sorry about the noise!

if users want to use topolvm they can manually enable/install it, we'll add a wiki entry with the steps
